src/python/pants/engine/internals/graph.py

```diff
@@ -168,85 +176,211 @@ def warn_deprecated_field_type(request: _WarnDeprecatedFieldRequest) -> _WarnDep
     return _WarnDeprecatedField()
 
 
+@dataclass(frozen=True)
+class _TargetParametrizations:
+    """All parametrizations and generated targets for a single input Address.
+
+    If a Target has been parametrized, it might _not_ be present in this output, due to it not being
+    addressable using its un-parameterized Address.
+    """
+
+    original_target: Target | None
+    parametrizations: FrozenDict[Address, Target]
+
+    def get(self, address: Address) -> Target | None:
+        if self.original_target and self.original_target.address == address:
+            return self.original_target
+        return self.parametrizations.get(address)
+
+    def generated_or_generator(self, maybe_generator: Address) -> Iterator[Target]:
+        if not self.original_target:
+            raise ValueError(
+                "A `parametrized` target cannot be consumed without its parameters specified.\n"
+                f"Target `{maybe_generator}` can be addressed as:\n"
+                f"{bullet_list(addr.spec for addr in self.parametrizations)}"
+            )
+        if self.parametrizations:
+            # Generated Targets.
+            yield from self.parametrizations.values()
+        else:
+            # Did not generate targets.
+            yield self.original_target
+
+
 @rule
-async def resolve_target(
+async def resolve_target_parametrizations(
     address: Address,
     registered_target_types: RegisteredTargetTypes,
     union_membership: UnionMembership,
     target_types_to_generate_requests: TargetTypesToGenerateTargetsRequests,
-) -> WrappedTarget:
-    if not address.is_generated_target:
-        target_adaptor = await Get(TargetAdaptor, Address, address)
-        target_type = registered_target_types.aliases_to_types.get(target_adaptor.type_alias, None)
-        if target_type is None:
-            raise UnrecognizedTargetTypeException(
-                target_adaptor.type_alias, registered_target_types, address
+    files_not_found_behavior: FilesNotFoundBehavior,
+) -> _TargetParametrizations:
+    assert not address.is_generated_target and not address.is_parametrized
+
+    target_adaptor = await Get(TargetAdaptor, Address, address)
+    target_type = registered_target_types.aliases_to_types.get(target_adaptor.type_alias, None)
+    if target_type is None:
+        raise UnrecognizedTargetTypeException(
+            target_adaptor.type_alias, registered_target_types, address
+        )
+    if (
+        target_type.deprecated_alias is not None
+        and target_type.deprecated_alias == target_adaptor.type_alias
+        and not address.is_generated_target
+    ):
+        await Get(_WarnDeprecatedTarget, _WarnDeprecatedTargetRequest(target_type))
+
+    target: Target | None
+    generate_request = target_types_to_generate_requests.request_for(target_type)
+    if generate_request:
```

Comment:
Sorry, I rushed this one in without doing this. I'm going to open a ticket for followup (since I'm out of town for the next few days), but will clean this up before `2.11.x`.